### PR TITLE
feat: combine download binary properties for image hash

### LINKED ISSUE #362: [CoAP simulator] Image not rebuild if the download URL (or its contents) changes
If the download URL (or the downloaded archive) is changed, the Docker image is not rebuild.

This means we cannot update the health check simulator with the latest release.

### PR DESCRIPTION
**Pull request by pudkrong**
## User Description:
Using `etag`, `content-length`, and `last-modified` of the CoAP simulator for change detection. So, system can rebuild the docker image.

### CODE REVIEW

File: cdk/backend.ts [hunk 1]
```diff
@@ -163,7 +163,22 @@ const coapDockerfilePath = path.join(
...
-const coapHash = await hashFolder(coapDockerfilePath)
+const coapSimulatorDownloadUrl = (
+	await getCoAPHealthCheckSettings({
+		ssm,
+		stackName: STACK_NAME,
+	})
+).simulatorDownloadURL.toString()
+const { headers, ok, status } = await fetch(coapSimulatorDownloadUrl, {
+	method: 'HEAD',
+})
```
Comment (coderbyheart) [inline cdk/backend.ts]: This needs to be cached, so it does not download the file every time we run CDK.

File: cdk/backend.ts [hunk 2]
```diff
@@ -163,7 +163,22 @@ const coapDockerfilePath = path.join(
...
+if (!ok) throw new Error(`Failed to download CoAP simulator: ${status}`)
+
+const coapSimulatorBinaryHash = `${headers.get('etag')}-${headers.get(
+	'content-length',
+)}-${headers.get('last-modified')}`
```
Comment (coderbyheart) [inline cdk/backend.ts]: Instead of using the server data, use the checksum of the file. Because that is what really matters and will tell if the contents have changed, not the metadata the webserver has.

**Review by coderbyheart — CHANGES_REQUESTED** [review]

File: cdk/backend.ts (hunk 1 above)
Comment (pudkrong) [reply cdk/backend.ts]: It does not download every time. Basically, I use `HEAD` to get file metadata. In S3, it provides `etag` in the header where it is never change if the file is the same.

File: cdk/backend.ts (hunk 2 above)
Comment (pudkrong) [reply cdk/backend.ts]: I did this at the first time, but we need to move the downloaded file to `cdk/resources/containers/coap` where it is docker context. Otherwise, docker won't allow you to copy the files to docker image. That's why I re-think and use server meta data to check.

Here is the more information about [S3 Etag](https://docs.aws.amazon.com/AmazonS3/latest/userguide/checking-object-integrity.html)

Comment (coderbyheart) [reply cdk/backend.ts]: You can download it to that location, no problem.

Comment (coderbyheart) [reply cdk/backend.ts]: This solution depends then on S3, which is OK, but needs to be documented. Because not every file host automatically provides etags.

File: aws/getOrBuildDockerImage.ts [hunk 3]
```diff
@@ -98,6 +102,7 @@ export const getOrBuildDockerImage =
 			debug?.(`Image tag ${imageTag} exists on ${repositoryUri}`)
 			return { imageTag }
 		} else {
+			if (beforeBuild !== undefined) await beforeBuild()
```
Comment (coderbyheart) [inline aws/getOrBuildDockerImage.ts]: Instead of inventing your own event messaging, use an EventEmitter. https://nodejs.org/api/events.html#class-eventemitter

File: cdk/backend.ts [hunk 4]
```diff
@@ -163,23 +168,46 @@ const coapDockerfilePath = path.join(
...
 	debug: debug('CoAP simulator docker image'),
+	beforeBuild: async () => {
+		await copyFile(
+			tmpCoapSimulatorPath,
+			path.join(coapDockerfilePath, 'coap-simulator.zip'),
+		)
+	},
+	afterBuild: async () => {
+		await rm(path.join(coapDockerfilePath, 'coap-simulator.zip'))
+	},
```
Comment (coderbyheart) [inline cdk/backend.ts]: Why is this hooked into the logic of getOrBuildDockerImage(), when `getOrBuildDockerImage` does nothing with these functions.

I'd like to see the logic separated to avoid this kind of workflow nesting.

1. Since we know based on the `coapHash` whether the Docker image needs to be rebuild (we can check whether the current docker image hash is the same as `coapHash`.
2. If not, download simulator ZIP and copy to expected file location for build (`copyeFile(...)`)
3. Build the new image
4. We don't need to delete `coap-simulator.zip`, so the `afterBuild` step can be removed.

File: docker/hashFolder.ts [hunk 5]
```diff
@@ -41,6 +41,14 @@ const generateMd5ForFolder = async (
 	return md5Hashes
 }
 
+export const hashStrings = (data: string[]): string => {
```
Comment (coderbyheart) [inline docker/hashFolder.ts]: Duplicate of https://github.com/hello-nrfcloud/backend/blob/99ca8ddd4059584fd4b5fcf084beec78276c50c6/cdk/helpers/lambdas/checksumOfFiles.ts#L15C14-L15C31

File: docker/hashFolder.ts [hunk 6]
```diff
@@ -2,7 +2,7 @@ import { createHash } from 'node:crypto'
 import { createReadStream, readdirSync, statSync } from 'node:fs'
 import { join } from 'node:path'
 
-const hashFile = async (file: string): Promise<string> => {
+export const hashFile = async (file: string): Promise<string> => {
```
Comment (coderbyheart) [inline docker/hashFolder.ts]: Duplicate of https://github.com/hello-nrfcloud/backend/blob/99ca8ddd4059584fd4b5fcf084beec78276c50c6/cdk/helpers/lambdas/checksumOfFiles.ts#L15C14-L15C31

**Review by coderbyheart — CHANGES_REQUESTED** [review]

File: aws/getOrBuildDockerImage.ts (hunk 3 above)
Comment (pudkrong) [reply aws/getOrBuildDockerImage.ts]: Event emitter cannot ensure the required file is successfully copied over.

File: cdk/backend.ts (hunk 4 above)
Comment (pudkrong) [reply cdk/backend.ts]: So, we have to execute function to check the existing tag twice, one before calling `getOrBuildDockerImage` and another one inside `getOrBuildDockerImage`.

Comment (pudkrong) [reply cdk/backend.ts]: Or we can copy file over no matter the docker will create a new image or not and delete it afterward to clean up. What do you think about this approach?

Comment (pudkrong) [reply cdk/backend.ts]: As per agreement, we will use `etag` to detect binary change and this will be documented in the README.

File: README.md [hunk 7]
```diff
@@ -28,7 +28,9 @@ employ in conjunction with Lambda to execute the CoAP simulator and transmit a
 message to nRF Cloud.
 
 To accomplish this, you need to supply the URL to the CoAP simulator binary in
-zip format via an SSM parameter for the stack.
+zip format via an SSM parameter for the stack. Please note that the server
+_MUST_ support `ETag` in order to detect binary changes and rebuild the Docker
```
Comment (coderbyheart) [inline README.md]:
```suggestion
_MUST_ serve a `HEAD` request with and `ETag` header containing the hash of the file
in order to detect binary changes and rebuild the Docker
```

File: cdk/backend.ts [hunk 8]
```diff
@@ -163,7 +164,25 @@ const coapDockerfilePath = path.join(
...
+const coapSimulatorDownloadUrl = (
+	await getCoAPHealthCheckSettings({
+		ssm,
+		stackName: STACK_NAME,
+	})
+).simulatorDownloadURL.toString()
+const { headers, ok, status } = await fetch(coapSimulatorDownloadUrl, {
+	method: 'HEAD',
+})
+if (!ok) throw new Error(`Failed to download CoAP simulator: ${status}`)
```
Comment (coderbyheart) [inline cdk/backend.ts]: status (a number) is not very helpful to understand the error. Print the message.

```suggestion
if (!ok) throw new Error(`Failed to download CoAP simulator: ${await res.text()}`)
```

File: cdk/backend.ts [hunk 9]
```diff
@@ -163,7 +164,25 @@ const coapDockerfilePath = path.join(
...
+const coapSimulatorBinaryHash = headers.get('etag')
+if (coapSimulatorBinaryHash === null)
+	throw new Error(`Server does not support ETag`)
```
Comment (coderbyheart) [inline cdk/backend.ts]:
```suggestion
	throw new Error(`[CoAP simulator] No ETag present in server response.`)
```

File: cdk/resources/containers/coap/Dockerfile [hunk 10]
```diff
@@ -20,8 +20,8 @@ RUN apt-get update && \
   tar
 
 # Download CoAP simulator binary
```
Comment (coderbyheart) [inline cdk/resources/containers/coap/Dockerfile]:
```suggestion
```

File: cdk/resources/containers/coap/src/handler.ts [hunk 11]
```diff
@@ -83,10 +83,14 @@ export const handler: Handler<Event, EventResponse> = async (event) => {
...
 		createDeviceProperties(devicePropertiesFile, event.deviceProperties)
 
+		const startTs = process.hrtime()
 		const result = await run({
 			command: 'coap-simulator/bin/coap-simulator',
 			args: event.args.concat(['-c', devicePropertiesFile]),
 		})
+		const [sec, ns] = process.hrtime(startTs)
+		const ms = sec * 1000 + ns / 1e6
+		console.log(`CoAP simulator: duration ${Math.round(ms)} ms`)
```
Comment (coderbyheart) [inline cdk/resources/containers/coap/src/handler.ts]: Unrelated change.

**Review by coderbyheart — CHANGES_REQUESTED** [review]
